nmigen/hdl/ir.py

```diff
@@ -532,6 +532,13 @@ def prepare(self, ports=None, missing_domain=lambda name: ClockDomain(name)):
         if ports is None:
             fragment._propagate_ports(ports=(), all_undef_as_ports=True)
         else:
+            # typecheck ports is either list or tuple (#362)
+            if not isinstance(ports, tuple) and not isinstance(ports, list):
```

Comment:
This isn't a major issue here, but I prefer to have typechecking code at the very beginning of the function so that it "fails fast" (before any other errors can arise).

Reply:
I'd like to keep this here, because it leverages the existing `if ports is None:` check. Otherwise we're checking for None redundantly (as we'd need that in the typecheck to avoid a false negative). I view this as being at the top of the "ports exists" branch. I don't feel _too_ strongly about this - I'll ultimately do what you say of course - but I wanted to explain the reasoning.

Reply:
Fine to keep as-is by me.